power: Add device idle power management support

ramakrishnapallala opened the pull request:
device: Extend device_set_power_state API to support async requests
power: Add device idle power management support
samples: power: Add test for device Idle PM

Commit:
device: Extend device_set_power_state API to support async requests

The existing device_set_power_state() API works only in synchronous
mode and this is not desirable for devices(ex: Gyro) which take
longer time (few 100 mSec) to suspend/resume.

To support async mode, a new callback argument is added to the API.
The device drivers can asynchronously suspend/resume and call the
callback function upon completion of the async request.

This commit adds the missing callback parameter to all the drivers
to make it compliant with the new API.

Signed-off-by: Ramakrishna Pallala <ramakrishna.pallala@intel.com>

ramakrishnapallala commented:
These changes are the outcome of the discussion in Dev Meeting related to PR: https://github.com/zephyrproject-rtos/zephyr/pull/12172

pizi-nordic commented on include/device.h:
```diff
@@ -208,6 +208,7 @@ extern "C" {
 
 struct device;
 
+typedef void (*device_pm_cb)(struct device *dev, int status);
```
Could we add user-defined context the the callback?

ramakrishnapallala replied:
Ok

pizi-nordic commented on include/device.h:
```diff
@@ -221,7 +222,7 @@ struct device_config {
 	int (*init)(struct device *device);
 #ifdef CONFIG_DEVICE_POWER_MANAGEMENT
 	int (*device_pm_control)(struct device *device, u32_t command,
-				 void *context);
```
What about something like this:
```
int (*device_pm_control)(struct device *device, u32_t command, void *arg, device_pm_cb cb, void *context)
typedef void (*device_pm_cb)(struct device *dev, int status, void *context);
```

ramakrishnapallala replied:
Ok

carlescufi replied:
addressed

ramakrishnapallala commented:
@pizi-nordic Can you look the PR once?

ramakrishnapallala commented:
@pizi-nordic  I have modified the device idle PM based in async device_set_power_state(). Please let me know if you any concerns with the design. Based on your feedback I will push the sample app.

ramakrishnapallala commented:
@pizi-nordic Can you please review the PR? based on your feedback i will work on sample app.

pizi-nordic commented on doc/reference/power_management/index.rst:
```diff
@@ -339,7 +339,7 @@ Device Set Power State
 .. code-block:: c
 
-   int device_set_power_state(struct device *device, u32_t device_power_state);
+   int device_set_power_state(struct device *device, u32_t device_power_state, device_pm_cb cb);
```
How I can pass custom data to the callback?
(We agreed on a slightly different API: https://github.com/zephyrproject-rtos/zephyr/pull/13382#discussion_r256769700)

ramakrishnapallala replied:
"int (*device_pm_control)(struct device *device, u32_t command, void *arg, device_pm_cb cb, void *context)
typedef void (*device_pm_cb)(struct device *dev, int status, void *context);"

I missed the arg parameter you mentioned in the comment.
How are planning to use arg parameter?

pizi-nordic replied:
The arg allows to pass some parameters specific for given operation. I not have exact scenario for its usage now, however many times I missed this functionality.

ramakrishnapallala replied:
OK sure, will update the PR accordingly.

pizi-nordic commented on doc/reference/power_management/index.rst:
```diff
@@ -339,7 +339,7 @@ Device Set Power State
 .. code-block:: c
 
-   int device_set_power_state(struct device *device, u32_t device_power_state);
+   int device_set_power_state(struct device *device, u32_t device_power_state, void *arg, device_pm_cb cb);
```
I suggest swapping cb and arg arguments. This would result in more convenient order.

ramakrishnapallala replied:
do you want to send the arg to callback() from driver set power function?

pizi-nordic replied:
Yes, This should be opaque value which is passed from device_set_power_state() caller to callback.

ramakrishnapallala replied:
Ok

pizi-nordic commented on subsys/power/device_pm.c:
```diff
@@ -0,0 +1,158 @@
+/*
```
This does not look like the FSM we agreed on during our call.
Do you need some kind of design proposal?

ramakrishnapallala replied:
Before I implement the FSM I wanted some clarity on the state transitions.

Consider the following scenario.
1. device idle requested suspend
2. pm_state change from ACTIVE -> SUSPENDING
3. driver tries to suspend the device but fails to do so
4. driver calls the callback with an error return code.
In this case, what should we set the device state to?

Also if the pm_state is in SUSPENDING and new request comes for SUSPEND what we should do? should we call the device_set_power_state() with new SUSPEND requests for return?

what action we should for new requests if the current case is SUSPENDING or RESUMING? we can not skip the new requests(ex: SUSPEND) even if it is the same as the previous (ex: SUSPENDING) request because the previous request might fail. It would be better to rely only on the actual state for queuing the requests.

This commit now uses the actual device PM states as pm_states so we don't need any translation of FSM to device states in the device_pm code. I feel this simplifies the whole logic.

ramakrishnapallala replied:
@pizi-nordic Please have look at my query above. Thanks.

pizi-nordic replied:
You cannot use actual device state because it not covers transition state.
Here is algorithm I have in my mind. I hope it explain all the cases you described.
As you see, the algorithm is trying to reduce number of calls to device_set_power_state().

1. Assumptions:
    - Device could change power state only when it is idle. There is no "forcing" power state, as this might render the system non-operational (for example shutting down SPI bus during write to SPI flash will corrupt data). Failure of setting power mode is a failure of device - you cannot longer rely on it.
    - Suspending and resuming operations cannot be terminated until complete.

2. Variables:
    - FSM state: (ACTIVE, SUSPENDED, SUSPENDING, RESUMING)
    - Target state: (ACTIVE, SUSPENDED).

3. Pseudocode:
```
device_suspend() {
	if (target_state == SUSPENDED) {
		/* Device will eventually reach the SUSPSENDED state */
		return;
	}

	target_state = SUSPENDED;
	trigger_fsm();
}

device_resume() {
	if (target_state == ACTIVE) {
		/* Device will eventually reach the ACTIVE state */
		return;
	}

	target_state = ACTIVE;
	trigger_fsm();
}

trigger_fsm() {
	run_through_workqueue(fsm);
}

/* As device_set_power_state() is asynchronous, this will no block workqueue */
fsm() {
	switch (fsm_state) {
		case SUSPENDING:
		case RESUMING:
			/* We have to wait */
			break;

		case ACTIVE:
			if (target_state == SUSPENDED) {
				device_set_power_state(SUSPENDED, trigger_fsm)
				fsm_state = SUSPENDING;
			}
			break;

		case SUSPENDED:
			if (target_state == ACTIVE) {
				device_set_power_state(ACTIVE, trigger_fsm)
				fsm_state = RESUMING;
			}
			break;
	}

	broadcast_state(fsm_state);
}

device_wait_suspended() {
	while (recv_broadcasted_state() != SUSPENDED) { sleep() }
}

device_wait_active() {
	while (recv_broadcasted_state() != ACTIVE) { sleep() }
}

device_suspend_sync() {
	device_suspend();
	device_wait_suspended();
}

device_resume_sync() {
	device_resume();
	device_wait_active();
}
```

This could be extended to cancel workqueue job if is is not longer needed, such shortcut must be carefully checked for race conditions. As you see multiple calls to suspend/resume are handled and the last request is valid for us.

ramakrishnapallala replied:
I will look into it and get back to you.

ramakrishnapallala replied:
>    * FSM state: (ACTIVE, SUSPENDED, SUSPENDING, RESUMING)
>    * Target state: (ACTIVE, SUSPENDED).

> device_suspend() {
> 	if (target_state == SUSPENDED) {
> 		/* Device will eventually reach the SUSPSENDED state */
> 		return;
> 	}
> 
> 	target_state = SUSPENDED;
> 	trigger_fsm();
> }
 
> device_resume() {
> 	if (target_state == ACTIVE) {
> 		/* Device will eventually reach the ACTIVE state */
> 		return;
> 	}
> 
> 	target_state = ACTIVE;
> 	trigger_fsm();
> }
> trigger_fsm() {
> 	run_through_workqueue(fsm);
> }
Are you suggesting we should have workquque in the device_pm.c? If so, do we need a saperate thread or we should submit the work on system thread?

> fsm() {
> 	switch (fsm_state) {
> 		case SUSPENDING:
> 		case RESUMING:
> 			/* We have to wait */
> 			break;
> 
> 		case ACTIVE:
> 			if (target_state == SUSPENDED) {
> 				device_set_power_state(SUSPENDED, trigger_fsm)
> 				fsm_state = SUSPENDING;
> 			}
> 			break;
> 
> 		case SUSPENDED:
> 			if (target_state == ACTIVE) {
> 				device_set_power_state(ACTIVE, trigger_fsm)
> 				fsm_state = RESUMING;
> 			}
> 			break;
> 	}
> 
> 	broadcast_state(fsm_state);
> }
is this the worked logic which maintains the states? also why broadcat fsm state?

pizi-nordic replied:
> is this the worked logic which maintains the states?

I am not fully getting this question. If you are asking which entity holds the device state: In the proposed model this is tracked by FSM.

> why broadcat fsm state?

Broadcasting the FSM state is needed for blocking version of the device_get/put. Whenever state is changed, a signal is sent. If FSM reached desired state, then you can exit from the blocking function.

Also, the application can track the device state and run some action if device reach desired state.

carlescufi replied:
addressed, at least when it comes to the system workqueue

ramakrishnapallala commented:
@pizi-nordic I have updated the PR as your comments. Please look at the overall flow/design of the code. I will update PR with sample app once you are fine with it.

pizi-nordic commented:
@ramakrishnapallala: OK. I will look at the code tomorrow morning.

ramakrishnapallala commented:
> @ramakrishnapallala: OK. I will look at the code tomorrow morning.

Gentle reminder.

pizi-nordic commented on subsys/power/device_pm.c:
```diff
@@ -0,0 +1,204 @@
+		k_work_init(&dev->pm->work, pm_work_handler);
+	} else {
+		if (dev->pm->usage) {
+			ret = device_set_power_state(dev,
```
This should go through FSM too. Managing state from single place will make the code simpler and easier to maintain.

ramakrishnapallala replied:
Ok

pizi-nordic reviewed (commented):
I apologize for the delay and thank you for the reminder :)

ramakrishnapallala commented on drivers/counter/counter_qmsi_aonpt.c:
```diff
@@ -196,7 +196,7 @@ static int aonpt_resume_device_from_suspend(struct device *dev)
 */
 static int aonpt_qmsi_device_ctrl(struct device *dev, u32_t ctrl_command,
-				  void *context)
+				  void *context, device_pm_cb cb, void *arg)
```
Yes.

carlescufi replied:
addressed

ramakrishnapallala commented on include/device.h:
```diff
@@ -208,6 +220,31 @@ extern "C" {
+	struct k_sem lock;
+	bool enable;
+	atomic_t usage;
+	u32_t fsm_state;
```
ok.

ramakrishnapallala commented on include/device.h:
```diff
@@ -208,6 +220,31 @@ extern "C" {
+ * @param signal signal to notify the Async API callers
+ */
+struct device_pm {
+	struct device *dev;
```
I will check once again.

carlescufi replied:
Did you find a way to avoid having a pointer to the device structure here?

ramakrishnapallala replied:
At present, I did not find any way.

ramakrishnapallala commented on include/device.h:
```diff
@@ -237,6 +274,9 @@ struct device {
 	const void *driver_api;
 	void *driver_data;
+#ifdef CONFIG_DEVICE_POWER_MANAGEMENT
+	struct device_pm *pm;
```
will check.

ramakrishnapallala commented on include/device.h:
```diff
@@ -447,6 +496,104 @@ int device_any_busy_check(void);
+#ifdef CONFIG_DEVICE_IDLE_PM
+
+/* Device PM FSM states */
+#define DEVICE_PM_FSM_ACTIVE			1
```
ok

ramakrishnapallala commented on include/device.h:
```diff
@@ -447,6 +496,104 @@ int device_any_busy_check(void);
+static inline void device_pm_disable(struct device *dev) { }
+static inline int device_pm_get(struct device *dev) { return 0; }
+static inline int device_pm_get_sync(struct device *dev) { return 0; }
+static inline int device_pm_put(struct device *dev) { return 0; }
```
Ok will look into it.

carlescufi replied:
Any follow-up here @ramakrishnapallala ?

ramakrishnapallala replied:
We would need to put the driver code under CONFIG_DEVICE_IDLE_PM macro as it does not make much sense to return false info.

ramakrishnapallala commented on kernel/device.c:
```diff
@@ -62,6 +62,9 @@ void _sys_device_do_config_level(s32_t level)
 			_k_object_init(info);
 		}
+#ifdef CONFIG_DEVICE_IDLE_PM
+		device_pm_enable(info);
```
sys_init? now we don't have sys_init in device_pm.c

pizi-nordic replied:
When you look at the SYS_INIT() macro you will see that it registers call using  DEVICE_AND_API_INIT(). The device_pm_enable() will be called on these registrations. I am not sure it this is safe.

ramakrishnapallala replied:
I initially placed this call in the driver init and moved to this place on your suggestion which made sense. I will double check the flow once.

pizi-nordic replied:
I remember that. Just recently I worked around device registration and the doubt raised in my mind.

ramakrishnapallala commented on samples/sensor/apds9960/src/main.c:
```diff
@@ -79,11 +79,11 @@ void main(void)
 
 		p_state = DEVICE_PM_LOW_POWER_STATE;
-		device_set_power_state(dev, p_state);
+		device_set_power_state(dev, p_state, NULL, NULL);
```
I think the driver should only call the callback if it is passed else it should simply ignore.

pizi-nordic replied:
Then how we can be sure that device is in desired state? (this is async call now).

ramakrishnapallala replied:
It's an application or driver responsibility to pass the callback, If it doesn't it may not get the desired result. Let it be this way for now.

pizi-nordic replied:
Ok. I personally would prefer assertion in such case.

carlescufi replied:
I agree, an __ASSERT() here would be safer

ramakrishnapallala commented on subsys/power/device_pm.c:
```diff
@@ -0,0 +1,204 @@
+
+static u32_t pm_to_fsm_state(u32_t pm_state)
+{
+	if (pm_state == DEVICE_PM_ACTIVE_STATE) {
```
device PM and FSM states are different except ACTIVE/SUSPENDED, it would be better we explicitly check and assign the corresponding values.

pizi-nordic replied:
Yes. But we can do that in declaration:

```
enum { A, B };
enum { C = A, D = B };

STATIC_ASSERT(A == C, B == D) /* in places where you are using casts */
```

Such solution will be smaller and lightweight.

ramakrishnapallala replied:
this would mean changing the power states to enums?

pizi-nordic replied:
Yes.

ramakrishnapallala commented on subsys/power/device_pm.c:
```diff
@@ -0,0 +1,204 @@
+	}
+}
+
+static void device_pm_callback(struct device *dev,
```
what should i use here? irq_lock?  we need to change the fsm_state right? else we need to make it atomic.

pizi-nordic replied:
This is the problem solved by the worqueue - there you have thread context and can use regilar locks. Do you see need for a lock for work scheduling?

ramakrishnapallala replied:
not for scheduling but if we are changing the fsm state we would need lock. I am working on moving most of the logic to worker now. let's see if we can avoid this

carlescufi replied:
addressed by workqueue

ramakrishnapallala commented on subsys/power/device_pm.c:
```diff
@@ -0,0 +1,204 @@
+			       int retval, void *context, void *arg)
+{
+	u32_t target_state = *((u32_t *)context);
+	u32_t fsm_state = *((u32_t *)arg);
```
device_set_power_state passes internally as pointer to device_pm_control().

The reason for passing these values is we can cross-check the response/callback is coming for the request which was triggered.

ramakrishnapallala commented on subsys/power/device_pm.c:
```diff
@@ -0,0 +1,204 @@
+
+static void pm_work_handler(struct k_work *work)
+{
+	struct device_pm *pm = CONTAINER_OF(work,
```
Ok.

carlescufi replied:
addressed

ramakrishnapallala commented on subsys/power/device_pm.c:
```diff
@@ -0,0 +1,204 @@
+	if (dev->pm->target_state == target_state) {
+		if ((pm_flags & DEVICE_PM_ASYNC) &&
+			(pm_to_fsm_state(target_state) ==
+				dev->pm->fsm_state)) {
```
are saying we should use the worker to handle all the FSM states?

pizi-nordic replied:
Yes. I think that will be best. Do you see any drawback of such solution?

ramakrishnapallala commented on subsys/power/device_pm.c:
```diff
@@ -0,0 +1,204 @@
+		} else {
+			k_poll_signal_check(&dev->pm->signal,
+						&signaled, &result);
+			if (result != target_state) {
```
ok will check.

pizi-nordic replied:
Still, you are waiting here for first signal, which might be not the you are waiting for.
Also, the signal state is not reset, so this will work only once.

pizi-nordic commented:
@ramakrishnapallala: We need to solve one more thing in the area:

The devices might have internal dependencies: For example the SPI device driver might use GPIO for a CS line and this GPIO might come from I2C expander. So the SPI driver in its  device_set_power_state(ACTIVE) handler have to call device_get(GPIO) and wait until the I2C expander becomes ready.  Then, when GPIO is available, it could use it communicate with the device and wake it up.

So far I do not see easy solution to this. We cannot use blocking version of device_get as this will result in blocking device_set_power_state() hander. Also we cannot wait for signal.

I am considering adding a callback to device_get(),  but this have several drawbacks or extending signal interface in order to be able to schedule a work when signal arrives.

@mnkp, @pdunaj: Do you see a simple and elegant solution to the problem described above?

pdunaj commented:
@pizi-nordic I was thinking about creating work woken up by sync objects (pollable objects). https://github.com/zephyrproject-rtos/zephyr/issues/10748

I think this solution would be great in Zephyr as it would make it possible to poll on various things without using a dedicated thread.

ramakrishnapallala commented:
> @pizi-nordic I was thinking about creating work woken up by sync objects (pollable objects). #10748
> 
> I think this solution would be great in Zephyr as it would make it possible to poll on various things without using a dedicated thread.

Ok this change can come in as an enhancement/improvement in different PR.

Commit:
power: Add device idle power management support

Add framework for device Idle Power Management(IPM)
for suspending devices based on device idle. This will
help in saving power even while system(CPU) is active.

The framework uses device_set_power_state() API set the
device power state accordingly based on the usage count.

Signed-off-by: Ramakrishna Pallala <ramakrishna.pallala@intel.com>

ramakrishnapallala commented:
@pizi-nordic  I have updated the PR, Please have a look.

- As of now, SYS_INIT does not harm anything.
 - I have not made the device PM states as enum, It's better we map/assign the states explicitly instead of implicit mapping.
- Also, we need to set the fsm state from callback else we would up in locking/worker scheduling related issues.

ramakrishnapallala commented:
@pizi-nordic I have updated the PR with the following changes:

- Moved device_pm_enable() to driver init. It was causing hang issues and causes unnecessary call for SYS _INIT.
- Introduced additional signal/event for internal use to signal the completion of PM request from the callback. With this now all the FSM state changes are happening from worker alone.

Note: As we are using system thread for queuing the work it's better to avoid nested calling of get/put API's from dependant driver's  pm_control(). The reason is if one blocks the thread (k_poll) the other cannot continue until the previous worker is done.

Also, the code freeze date around the corner, can please you prioritize this PR review.

pizi-nordic commented on subsys/power/device_pm.c:
```diff
@@ -0,0 +1,243 @@
+	}
+
+	/* Wait for the device callback singnal event */
+	ret = k_poll(&dev->config->pm->i_event, 1, K_FOREVER);
```
Here you are blocking the whole workqueue. This is not acceptable. Please look again at the pseudocode I prepared.

ramakrishnapallala replied:
removed the k_poll from worker.

pizi-nordic reviewed (changes requested):
Blocking worqueue is not acceptable. Also, please review commends and documentation. It looks that there are few copy/paste errors.

ramakrishnapallala commented on include/device.h:
```diff
@@ -208,6 +225,36 @@ extern "C" {
+	struct k_poll_event i_event;
+	struct k_poll_signal i_signal;
+	struct k_poll_event event;
+	struct k_poll_signal signal;
```
will remove the k_poll stuff. Later we can look at #10748

carlescufi replied:
This is still fairly heavy on memory, @ramakrishnapallala do you have plans to address this this week?

ramakrishnapallala replied:
This is the best we can have now.

ramakrishnapallala commented on include/device.h:
```diff
@@ -447,6 +504,106 @@ int device_any_busy_check(void);
+
+/* Device PM FSM states */
+enum device_pm_fsm_state {
+	DEVICE_PM_FSM_ACTIVE = 1,
```
will fix it

ramakrishnapallala commented on include/device.h:
```diff
@@ -447,6 +504,106 @@ int device_any_busy_check(void);
+ *
+ * Called by a device driver to enable device idle power management.
+ *
+ * @param dev Pointer to device structure of the specific device driver
```
Ok

carlescufi replied:
addressed

ramakrishnapallala commented on samples/subsys/power/device_pm/src/dummy_driver.c:
```diff
@@ -0,0 +1,169 @@
+		return ret;
+	}
+
+	ret = device_pm_get(dev);
```
My main intention was to use the API's from drivers itself but they are not limited to drivers.

ramakrishnapallala commented:
> Blocking worqueue is not acceptable. Also, please review commends and documentation. It looks that there are few copy/paste errors.

@pizi-nordic I have addressed all the comments. Can you please review once?

ramakrishnapallala commented:
Reminder: @pizi-nordic

pizi-nordic commented on subsys/power/device_pm.c:
```diff
@@ -0,0 +1,226 @@
+	bool set_power = false;
+	u8_t target_state;
+
+	k_sem_take(&dev->config->pm->lock, K_FOREVER);
```
This could block, stopping whole worqueue.

pizi-nordic reviewed (commented):
There are still places where we can stop execution waiting for something. I think, that this excessive locking could be eliminated as well as the size of data needed to manage the device.

Please note the following:
- The target state, is basically equal to usage count: If usage > 0, our target state is ACTIVE, if usage == 0, our target state is SUSPENDED. 
- There will be no situation when the FSM will compete with itself, as workqueue serializes all the requests. So there is no locking needed. there.

So device_get() / device_put() could be reduced to:
```
/* Pseudocode */
device_get()
{
if (atomic_inc(usage) == 0)
    k_work_submit(fsm_work);
}

device_put()
{
if (atomic_dec(usage) == 1)
    k_work_submit(fsm_work);
}
```
The k_work_submit() will take care about serialization (see https://github.com/zephyrproject-rtos/zephyr/pull/13382#discussion_r262047296) and the atomic_inc()/dec() serializes usage update.

The device_pm_callback is also very simple:
```
device_pm_callback(void *context)
{
atomic_set(&fsm_state, (fsm_state_t)(context));
k_work_submit(fsm_work);
}
```

Matching FSM will look like this:
```
fsm_work()
{
state = atomic_get(&fsm_state);
switch (state)
{
    case ACTIVE:
           if (atomic_get(&usage) == 0) {
                 atomic_set(&fsm_state, SUSPENDING)
                 device_set_power_state(...., device_pm_callback, SUSPENDED);
           } else {
           /* Send signal to notify that device is active. */
           }
           break;
    case SUSPENDED:
            /* In this state the FSM owns the device. No other FSM instance nor device_pm_callback will execute */
        if (atomic_get(&usage) != 0) {
                 atomic_set(&fsm_state, RESUMING)
                 device_set_power_state(...., device_pm_callback, ACTIVE);
           } else {
           /* Send signal to notify that device is suspended. */
           }
          break'

    default:
        /* Do nothing: We are waiting for  device_pm_callback(); */
       break;
}
```

ramakrishnapallala commented on subsys/power/device_pm.c:
```diff
@@ -0,0 +1,226 @@
+	}
+
+	/* Lock only if callback is called asynchronously */
+	k_sem_take(&dev->config->pm->lock, K_FOREVER);
```
I see one issue here with logic callback. 

device_pm_callback(void *context)
{
atomic_set(&fsm_state, (fsm_state_t)(context));
k_work_submit(fsm_work);
}

After callback when the worker gets called, the loop will again run forever because if we set the fsm to active/suspended in callback then again in the worker will run through the same logic.
we would need an additional variable (target state) to know the actual device state otherwise we would run through the loop continuously. 
this would mean we would need another atomic variable for target state?

ramakrishnapallala replied:
I realized we are using the usage count instead of target_state in the worker, but how do we handle device_set_power failure cases where the callback is called error code.

ramakrishnapallala replied:
@pizi-nordic Gentle reminder...

pizi-nordic replied:
> I see one issue here with logic callback.
> (...)
> After callback when the worker gets called, the loop will again run forever because if we set the fsm to active/suspended in callback then again in the worker will run through the same logic.

What loop? I do not see any loop in the proposed design.
- In the device_pm_callback you set the FSM state to either SUSPENDED to ACTIVE.
- In the work you are checking the FSM. If you are in SUSPENDED or ACTIVE state **and** only if the given state is not the desired one (obtained from the usage count), you are triggering next transition.

ramakrishnapallala replied:
I mentioned the loop w.r.t using target_state. If we use usage count we will not have loop the loop issue.

My concern was with error handling and with device_pm_enable/disable(). These are solved with use of target_state  atomic variable.

pizi-nordic replied:
>  I realized that we are using the usage count instead of target_state in the worker, but how do we handle device_set_power failure cases where the callback is called error code.

IMHO error here is a critical bug and we cannot continue. We can have 2 source of errors:
- Invalid state of the device (for example when we try suspending busy device): This is usage error. If the user use the device_get/put API correctly we will never run into this.
- I/O error (like I2C bus communication error or something similar): This is unrecoverable error at the moment. We cannot be sure what is real state of peripherar or even if we are still able to communicate with it. If there will be API to reset device, we would be able to use it to regain control over the device, but this will only delay the problem (what if reset fails?).

ramakrishnapallala replied:
So how do we handle the error case?

- we can either do ASSERT on error
- Or log the error and set the device to the previous state. If we set fsm state to the previous state it may not match the usage count, in this case, the loop will run forever if the callback return error again and again.

ramakrishnapallala commented:
@pizi-nordic  I have updated the PR with the following changes, Please have a look.

- Made fsm_state and target_state as atomic variables as we don't want to block the thread by using sem_lock.
- We cannot rely on the usage count variable in the worker because we will trigger the worker from device_pm_enable/disable API's. So we need to use target_state in the worker. Also, target_state can be used to check if there is any new request before submitting the worker.

pizi-nordic commented:
> We cannot rely on the usage count variable in the worker because we will trigger the worker from device_pm_enable/disable API's. So we need to use target_state in the worker. Also, target_state can be used to check if there is any new request before submitting the worker.

I am not getting this. Why we cannot use usage count?
Could you please give me example when relaying on the usage will fail?

I assume that you have in mind the situation similar to this one:
- Initial state: Device suspended, usage count = 0
- Caller A: device_get() => (usage += 1) = 1
- Caller B: device_put() => (usage -= 1) = 0;
- FSM work executed (usage = 0, device suspended): No operation.

- We can tread called A and B as the same entity. In such case device_put() cancels the previous device_get(). If the device not started wake-up procedure, it will be abandoned, otherwise full cycle (SUSPENDED => ACTIVE => SUSPENDED) will be perfomed.

- If the called A and B are fully independent, this is usage error as put and get were not paired at some time, which resulted in this situation.
Inm

ramakrishnapallala commented:
> > We cannot rely on the usage count variable in the worker because we will trigger the worker from device_pm_enable/disable API's. So we need to use target_state in the worker. Also, target_state can be used to check if there is any new request before submitting the worker.
> 
> I am not getting this. Why we cannot use usage count?
> Could you please give me example when relaying on the usage will fail?
> 
> I assume that you have in mind the situation similar to this one:
> 
> * Initial state: Device suspended, usage count = 0
> * Caller A: device_get() => (usage += 1) = 1
> * Caller B: device_put() => (usage -= 1) = 0;
> * FSM work executed (usage = 0, device suspended): No operation.
> * We can tread called A and B as the same entity. In such case device_put() cancels the previous device_get(). If the device not started wake-up procedure, it will be abandoned, otherwise full cycle (SUSPENDED => ACTIVE => SUSPENDED) will be perfomed.
> * If the called A and B are fully independent, this is usage error as put and get were not paired at some time, which resulted in this situation.
>   Inm

Say usage count is 0, and someone calls device_pm_disable(), in this case, we should resume the driver but if the worker checks usage count it will simply pass through as the device is already suspended whereas our intention is to resume the device.  For this purpose, we can use target_state instead of usage count.

ramakrishnapallala commented:
@pizi-nordic to summarize, we have following opens w.r.t review:

Need for target_state variable and why usage count may not work:
Say usage count is 0, and someone calls device_pm_disable(), in this case, we should resume the driver but if the worker checks usage count it will simply pass through as the device is already suspended whereas our intention is to resume the device. For this purpose, we can use target_state instead of usage count.

So how do we handle the error case in the callback()?
- we can either do ASSERT on error
- Or log the error and set the device to the previous state. If we set fsm state to the previous state it may not match the usage count, in this case, the loop will run forever if the callback return error again and again.

Let me know your feedback so that we can close on this PR.

pizi-nordic commented:
>Need for target_state variable and why usage count may not work:
Say usage count is 0, and someone calls device_pm_disable(), in this case, we should resume the driver but if the worker checks usage count it will simply pass through as the device is already suspended whereas our intention is to resume the device. For this purpose, we can use target_state instead of usage count.

IMHO handling that is simple. You only have to change conditions in the FSM:
```
(...)
    case ACTIVE:
           if (atomic_get(&usage) == 0 && pm_enabled == true) {
                 atomic_set(&fsm_state, SUSPENDING)
                 device_set_power_state(...., device_pm_callback, SUSPENDED);
           } else {
           /* Send signal to notify that device is active. */
           }
           break;
    case SUSPENDED:
            /* In this state the FSM owns the device. No other FSM instance nor device_pm_callback will execute */
        if (atomic_get(&usage) != 0 || pm_enabled == false) {
                 atomic_set(&fsm_state, RESUMING)
                 device_set_power_state(...., device_pm_callback, ACTIVE);
           } else {
           /* Send signal to notify that device is suspended. */
           }
          break;
(...)
```

The  pm_enable/pm_disable functions will look similarly to the get/put - they will set the flag accordingly and trigger FSM. Atomic is not needed as we are only setting/getting the value of the flag.


> So how do we handle the error case in the callback()?
>-   we can either do ASSERT on error
>-   Or log the error and set the device to the previous state. If we set fsm state to the previous state it may not match the usage count, in this case, the loop will run forever if the callback return error again and again.

I think that the __ASSERT is the sole solution here. The other approach will not work as we have no idea about the state of the hardware/driver if error was signalled. For example the device might be stuck somewhere in the middle of state change and will not work as expected until reset.

ramakrishnapallala commented:
@pizi-nordic I have addressed the comments as per your suggestions. Please review the PR.

pizi-nordic commented on subsys/power/device_pm.c:
```diff
@@ -0,0 +1,194 @@
+	k_sem_take(&dev->config->pm->lock, K_FOREVER);
+	/* Bring up the device before disabling the Idle PM */
+	k_work_submit(&dev->config->pm->work);
+	dev->config->pm->enable = false;
```
I suggest moving `dev->config->pm->enable = false;` before `k_work_submit()`. If work queue thread has higher priority than caller thread, the work will be executed before enable flag change and the FSM will not update device state.

ramakrishnapallala replied:
wil fix it.

pizi-nordic reviewed (commented):
Looks good from high level. I started detailed review (see my comments), but this is not finished yet. I will continue tomorrow.

ramakrishnapallala commented:
@pizi-nordic I have updated the PR, please have a look. Thanks.

pizi-nordic commented on drivers/counter/counter_qmsi_aonpt.c:
```diff
@@ -196,20 +196,25 @@ static int aonpt_resume_device_from_suspend(struct device *dev)
+		cb(dev, ret, context, arg);
+	}
+
+	return ret;
```
We have to differentiate the device_set_power_state() return value and the return code passed to callback:
- If device_set_power_state() return something other that 0, then callback will not be called and operation failed.
- If device_set_power_state() return 0, the operation status is reported through callback which will be/was called.

Also, this has to be documented.

ramakrishnapallala replied:
Ok

carlescufi replied:
addressed

pizi-nordic reviewed (commented):
Looks really good. Needs only some polish to be perfect :)

mnkp commented:
I have a question regarding the first commit "device: Extend device_set_power_state API to support async requests". The commit message says:
```
The existing device_set_power_state() API works only in synchronous
mode and this is not desirable for devices(ex: Gyro) which take
longer time (few 100 mSec) to suspend/resume.
```
However, the mere fact that some devices may take long time to suspend/resume does not seem like a good enough reason to make `device_set_power_state` asynchronous. If I need to wait for a device to wake up then I can wait in a thread I'm calling the function from. It's after all what the threads are made for. The rest of the system is running and if there is anything to do in a meantime it should be done elsewhere. I have a trouble coming up with a scenario when I would like the thread to proceed before the device is fully suspended (or resumed).

Looking at the remaining implementation let's say I have an external audio chip, or the aforementioned gyro and the chip documentation says that after waking the device up by toggling a gpio pin I need to wait 10 ms for the internal power to stabilize before the device can be considered active. The chip does not provide any interrupt to let me know that it's active. I simply need to wait 10 ms. The implementation of the `device_set_power_state()` function must be non blocking as it's submitted on the system work queue. So how exactly do I wait 10 ms? I cannot call `k_sleep()` from within `device_set_power_state()` as this would be blocking (for the system workqueue). Am I supposed to use a kernel timer? These all sounds fairly complicated.

My question is: what kind of real life use case are we trying to solve that requires an async version of `device_set_power_state()` function that cannot be solved by its synchronous version + the standard thread support?

The implementation proposed in this PR is complex and seems to go against the very principles Zephyr is based on. According to the [Synchronous Calls](https://docs.zephyrproject.org/latest/devices/drivers/drivers.html#synchronous-calls) paragraph of the Device Drivers and Device Model:
"High-level calls accessed through device-specific APIs, such as i2c.h or spi.h, are usually intended as synchronous. Thus, these calls should be blocking."

ramakrishnapallala commented on subsys/power/device_pm.c:
```diff
@@ -0,0 +1,185 @@
+LOG_MODULE_DECLARE(power);
+
+/* Device PM request type */
+#define DEVICE_PM_ASYNC			(1 << 0)
```
Ok will add the macro.

carlescufi replied:
addressed

ramakrishnapallala commented on subsys/power/device_pm.c:
```diff
@@ -0,0 +1,185 @@
+			"Invalid device PM state requested");
+
+	if (target_state == DEVICE_PM_ACTIVE_STATE) {
+		if (atomic_inc(&dev->config->pm->usage) < 0) {
```
this will avoid code duplication but anyways will check it again.

ramakrishnapallala commented on subsys/power/device_pm.c:
```diff
@@ -0,0 +1,185 @@
+			"Invalid device PM state requested");
+
+	if (target_state == DEVICE_PM_ACTIVE_STATE) {
+		if (atomic_inc(&dev->config->pm->usage) < 0) {
```
this is just to make sure we are also considering the case where some app/driver might call put() multiple times without calling get().

ramakrishnapallala commented on subsys/power/device_pm.c:
```diff
@@ -0,0 +1,185 @@
+	k_work_submit(&dev->config->pm->work);
+
+	/* Incase of Sync request wait for completion event */
+	if (!(pm_flags & DEVICE_PM_ASYNC)) {
```
Ok.

ramakrishnapallala commented on subsys/power/device_pm.c:
```diff
@@ -0,0 +1,185 @@
+	}
+
+	if (ret < 0) {
+		LOG_ERR("Set Power state error(%d)\n", ret);
```
Ok

ramakrishnapallala commented on subsys/power/device_pm.c:
```diff
@@ -0,0 +1,185 @@
+
+fsm_out:
+	k_poll_signal_raise(&dev->config->pm->signal, pm_state);
+
```
OK

ramakrishnapallala commented on subsys/power/device_pm.c:
```diff
@@ -0,0 +1,185 @@
+	/* Set the fsm_state */
+	if (*((u32_t *)context) == DEVICE_PM_ACTIVE_STATE) {
+		atomic_set(&dev->config->pm->fsm_state,
+				DEVICE_PM_FSM_STATE_ACTIVE);
```
then again I will have to add temp local variable for storing target fsm state. Can we leave this as it for now.

ramakrishnapallala commented on samples/subsys/power/device_pm/src/dummy_driver.c:
```diff
@@ -0,0 +1,173 @@
+			ret = dummy_suspend(dev);
+		}
+		break;
+	case DEVICE_PM_GET_POWER_STATE:
```
Yes.

ramakrishnapallala commented on subsys/power/device_pm.c:
```diff
@@ -0,0 +1,185 @@
+	k_sem_take(&dev->config->pm->lock, K_FOREVER);
+	dev->config->pm->enable = true;
+
+	/* During the driver init, device can set the
```
Adding in the DEVICE_DEFINE macro description.

carlescufi replied:
addressed

Commit:
samples: power: Add test for device Idle PM

Added test for Device Idle Power Management to invoke
device_pm_get/put API's.

Signed-off-by: Ramakrishna Pallala <ramakrishna.pallala@intel.com>

ramakrishnapallala commented:
recheck

ramakrishnapallala commented:
@pizi-nordic please review the updated the PR.

ramakrishnapallala commented:
recheck

carlescufi commented on include/device.h:
```diff
@@ -447,6 +504,106 @@ int device_any_busy_check(void);
+ * @brief Disable device idle PM
+ *
+ * Called by a device driver to disable device idle power management.
+ *
```
Addressed

carlescufi reviewed (changes requested):
Most of the comments from @pizi-nordic seem to have been addressed, with some notable exceptions regarding memory usage when it comes to what gets stored in the pm structure. I will wait for another round of feedback from @ramakrishnapallala and then we can take a decision for code freeze.

ramakrishnapallala commented:
@pizi-nordic @carlescufi I replied to your comments, can you check.

carlescufi reviewed (approved):
Approving with reservations due to the currently unresolved issues. I believe @pizi-nordic already owns a number of GitHub issues related to improving power management in general, I think we could add one more about device idle improvements.
